v03_pipeline/lib/methods/sex_check.py

```diff
@@ -0,0 +1,207 @@
+from __future__ import annotations
+import io
+from enum import Enum
+
+import hail as hl
+import matplotlib.pyplot as plt
+
+from v03_pipeline.lib.model import ReferenceGenome
+
+IMPUTE_SEX_ANNOTATIONS = [
+    'is_female',
+    'f_stat',
+    'n_called',
+    'expected_homs',
+    'observed_homs',
+    'sex',
+]
+
+
+class Ploidy(Enum):
+    AMBIGUOUS = 'ambiguous_sex'
+    ANEUPLOIDY = 'sex_aneuploidy'
+    FEMALE = 'XX'
+    MALE = 'XY'
+
+
+def get_contig_cov(
+    mt: hl.MatrixTable,
+    reference_genome: ReferenceGenome,
+    contig: str,
+    call_rate_threshold: float,
+    af_threshold: float = 0.01,
+) -> hl.Table:
+    """
+    Calculate mean contig coverage.
+
+    :param mt: MatrixTable containing samples with chrY variants
+    :param reference_genome: ReferenceGenome, either GRCh37 or GRCh38
+    :param contig: Chosen chromosome.
+    :param call_rate_threshold: Minimum call rate threshold.
+    :param af_threshold: Minimum allele frequency threshold. Default is 0.01
+    :return: Table annotated with mean coverage of specified chromosome
+    """
+    valid_contigs = {*reference_genome.autosomes, *reference_genome.sex_chromosomes}
+    if contig not in valid_contigs:
+        msg = f'Contig: {contig} is invalid for this reference genome'
+        raise ValueError(msg)
+    mt = hl.filter_intervals(
+        mt,
+        [hl.parse_locus_interval(contig, reference_genome=reference_genome.value)],
+    )
+    if contig == hl.get_reference(reference_genome.value).x_contigs[0]:
+        mt = mt.filter_rows(mt.locus.in_x_nonpar())
+    if contig == hl.get_reference(reference_genome.value).y_contigs[0]:
+        mt = mt.filter_rows(mt.locus.in_y_nonpar())
+
+    # Filter to common SNVs above defined callrate (should only have one index in the array because the MT only contains biallelic variants)
+    mt = mt.filter_rows(mt.AF > af_threshold)  # noqa: SIM300
+    mt = hl.variant_qc(mt)
+    mt = mt.filter_rows(mt.variant_qc.call_rate > call_rate_threshold)
+    mt = mt.select_cols(**{f'{contig}_mean_dp': hl.agg.mean(mt.DP)})
+    return mt.cols()
+
+
+def generate_fstat_plot(
+    ht: hl.Table,
+    xy_fstat_threshold: float,
+    xx_fstat_threshold: float,
+) -> io.BytesIO:
+    # Plot histogram of fstat values
+    # Returns the plot saved as a binary buffer.
+    buf = io.BytesIO()
+    df = ht.to_pandas()
+    plt.clf()
+    plt.hist(df['f_stat'])
+    plt.xlabel('Fstat')
+    plt.ylabel('Frequency')
+    plt.axvline(xy_fstat_threshold, color='blue', linestyle='dashed', linewidth=1)
+    plt.axvline(xx_fstat_threshold, color='red', linestyle='dashed', linewidth=1)
+    plt.savefig(buf, format='png')
+    buf.seek(0)
+    return buf
```

Comment:
This method returns a binary buffer now rather than writing to a static location.  the tasks themselves will manage paths/writing.

Reply:
what is the utility of a plot? It seems unlikely that this will be able to feed into any sort of automated decision making

Reply:
I think we've been using it as a validation step to ensure the `impute_sex` algorithm hasn't behaved incorrectly and requires human intervention.  

My hope is that, if the image is actually used, we can post it to slack to get a human "OK" to proceed with the DAG.  

@matren395 and @mike-w-wilson would you mind providing some context about the plot and how often it's needed?

Reply:
> My hope is that, if the image is actually used, we can post it to slack to get a human "OK" to proceed with the DAG.

When @mike-w-wilson and I initially discussed automating this workflow, the rough flow we had agreed on is that this check would remove failed families from the subset to load and then proceed with loading for all passed families in the project. The failed families would be reported back to the PMs somehow, and then they can always re-trigger loading a project whenever they felt like relatedness had been resolved. If this plot would be valuable to the PMs, I could see the utility in generating it for failed samples, but would want to make sure its something they can intepret

Reply:
The manual review of the plot is to confirm  the cutoff used for the imputed chromosomal sex label is a good cutoff. It is not a check of provided information vs imputed. The plot is bimodal so its a very quick check. However, since we started running this, I have never adjusted the cutoff threshold and I dont expect it to change for our internal callsets given the samples are the same and the gvcfs havent changed. But it would be appropriate to check the plot whenever we there is a major change in the DSPs pipelines, so if our data is being recalled with the GVS pipeline or gVCFs start getting generated with DRAGEN. The cutoff will likely stay the same but this would ensure they arent calling more homozygotes than we are now.

Reply:
In that case I think this function belong in a methods repo, not the seqr loading pipeline repo

Reply:
+1 

Per convo with @mike-w-wilson we're going to remove the plot and be more aggressive about moving parameters to constant.  

I'm also going to spend a little bit of time trying to brainstorm a heuristic we could use to trigger a failure if the plot doesn't show a clear bimodal.

Reply:
I've cleaned this up (and simplified) quite a bit.

1) Removed the "contig_coverage" logic.  Per @mike-w-wilson that was experimental and isn't used in practice; `hl.impute_sex` is going to be sufficient for what we need.
2) Added a simple heuristic to validate the `f_stat` distribution.  "If more than 1/100th of samples have `f_stats` between the `0.5` and `0.75` bounds and are classified as ambiguous, throw an error".  If 1/100 is too lenient we can lower the threshold to something safer!